pkg/kv/kvserver/kvflowcontrol/replica_rac2/admission.go

```diff
@@ -0,0 +1,277 @@
+// handleRaftReady when processing the corresponding MsgStorageAppend.
+//
+// Maintaining accuracy in the presence of an existing leader doing retransmits
+// is tricky, since the caller does not know which MsgApps were accepted by
+// Raft during a step. We can have two kinds of situations:
+//
+//   - [C1] e10 was stepped, but not accepted, since e9 was missing. e10 is
+//     stepped successfully later.
+//
+//   - [C2] e10 was successfully stepped, but the leader did not realize it, and
+//     later resends e10 that is ignored.
+//
+// We specify a weak correctness requirement: if a leader has not overridden
+// to LowPri, the lowPriOverrideState should not think it has. This prevents a
+// situation where an entry is consuming regular tokens on the leader, but
+// will be admitted as elastic work on the follower.
+//
+// That is, we allow for the following error: the leader has overridden to
+// LowPri, but lowPriOverrideState thinks it has not.
+//
+// RACv2 has an invariant that if a leader uses LowPri override once for an
+// entry, it must be in the send-queue, and will never again be used without
+// an override. This gives us an easy way to achieve the correctness
+// requirement -- use the specification of override (or lack thereof) from the
+// first time that a leader provides us information for an index, and don't
+// change it.
+//
+// For C2, consider the following sequence:
+// - e10 NormalPri stepped successfully
+// - e10 LowPri override not stepped successfully
+// - e10 emerges in handleRaftReady: we will use NormalPri
+//
+// For C1, consider the following sequence:
+//   - e10 NormalPri not stepped successfully
+//   - e10 LowPri override stepped successfully
+//   - e10 emerges in handleRaftReady: we will use NormalPri, though should have
+//     used LowPri.
+//
+// TODO(sumeer): consider increasing the accuracy by returning information
+// from step.
+type lowPriOverrideState struct {
+	// intervals for which information has been provided.
+	//
+	// A call to getEffectivePriority for index i causes a prefix of indices <=
+	// i to be discarded.
+	intervals []interval
+	// Highest term observed so far.
+	leaderTerm uint64
+}
+
+// Represents [first, last].
+type interval struct {
+	first          uint64
+	last           uint64
+	lowPriOverride bool
+}
+
+// sideChannelForLowPriOverride is called on follower replicas, and specifies
+// information for [first, last]. An index can never be queried using
+// getEffectivePriority without it being first specified in
+// sideChannelForLowPriOverride.
+//
+// A new leader can overwrite previous log entries. And the existing leader
+// can retransmit entries, with a LowPri override. In the case of an existing
+// leader we only extend the existing state, as discussed in the correctness
+// comment above.
+//
+// INVARIANT: first <= last.
+func (p *lowPriOverrideState) sideChannelForLowPriOverride(
+	leaderTerm uint64, first, last uint64, lowPriOverride bool,
```

Comment:
Do we need to explicitly record intervals with override == false?

I think we can remove the `lowPriOverride bool` from `interval`, and just add a `lastIndex` into `lowPriOverrideState`. All intervals explicitly present in the data structure signify an override, and all other entries up to `lastIndex` have no override.